### PR TITLE
46272784-Siamese-Network

### PR DESCRIPTION
**Pull request by PDU0**
This is a pull request to complete my submission for COMP3710's final project. In this project I implimented a siamese network to classify Alzheimer’s disease. The model is trained using ADNI brain image dataset. The training and validation accuracy are all above 90%.
For more information, please read the README.md file

Haolong Du (46272784)

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
- Commit messages are OK
- Readme is not satisfactory. Not much details on the design process, experiments or the training process, and no discussion on the results
- Loss of the network is not discussed in the readme. I'm assuming the accuracy plot shown is for classification, but no mention of how the input to the classification model looks like in the readme (pairs of images or embeddings of individual images?) Looking at the code, input to the entire model seems to be 2D pairs
- Performance is not satisfactory. Seems to be overfitting, but no other metrics or prediction results are shown for a better judgement

**Comment by PDU0** [comment]
Hello Nathasha,

Apologies for the lack of details in the README, this is my first time writing a README file and I was a bit confused about how many details should be put in README. 

1. According to requirement 1 (Section 6, task sheet), the description of the algorithm and how the algorithm works should be approximately one paragraph each, which leads me to believe that only a brief description is needed. Also, it appears that none of the requirements stated that a discussion on the results is needed.

2. The loss function used for fraining the Siamese network is Binary cross entropy, which can be found in the loss() function in modules.py. Sorry for not mentioning it.
The input of the siamese network is indeed a pair of images and the label (0 or 1). For classification (predict() function in predict.py), the input contains a test image (unknown class) and a list of example images [ADimg, NCimg], the test image is then compared with each of the example images using the Siamese network (the output of the Siamese is the similarity score between the test and the example image). If the similarity score between test image and AD example image is higher than the similarity score between test image and NC example, then the test image is classified as AD, and vice versa.

3. The function random_predict_one_shot() in predict.py performs 600 random one-shot classifications and returns the performance (classification accuracy). The accuracy is between 62% to 68%, which is indeed not satisdatory since it is below the 80% mark specified on the task sheet. This clearly shows the model overfits since the testing and validation accuracy is well above 80%. 

Once again, sorry for not putting more detail in the README.

**Comment by shakes76** [comment]
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate use and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem, some overfitting observed, could have attempted to fix -1
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage 
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages (most commit message could be more informative) -1
Progressive commits used
## Documentation
ReadMe OK, could use more info on loss, intro etc. -2
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description OK, could be more informative -1
